[Lens] Add metric Viz config options, title position and sizing

shahzad31 opened the pull request:
## Summary

Adds options to set title position and sizing in the metric viz

i am hoping this makes sense, we need to use this User experience app. 

![image](https://user-images.githubusercontent.com/3505601/154532835-d2dbd126-cb8c-4f4f-b775-62230dac9931.png)

Potential use cases 

![image](https://user-images.githubusercontent.com/3505601/151819953-707382fb-a546-4def-980a-26ccf16fcc81.png)

ghudgins commented:
related to https://github.com/elastic/kibana/issues/120386

markov00 commented on x-pack/plugins/lens/public/metric_visualization/metric_config_panel/title_aize_option.tsx:
```diff
@@ -0,0 +1,58 @@
+  setState: (newState: MetricState) => void;
+}
+
+const titleSizes = [
+  { id: 'xs', label: 'XS' },
+  { id: 's', label: 'S' },
+  { id: 'm', label: 'M' },
+  { id: 'l', label: 'L' },
+  { id: 'xl', label: 'XL' },
+  { id: 'xxl', label: 'XXL' },
+];
```
EUI has a `base` size that is between the `m` and `l` https://eui.elastic.co/#/theming/sizing
I think we should align to that as well since the CSS is using the `eui` multipliers

markov00 commented on x-pack/plugins/lens/public/metric_visualization/expression.scss:
```diff
@@ -11,11 +11,81 @@
+      }
+    }
+    &.titleSizeS {
+      .lnsMetricExpression__title {
+        font-size: $euiSizeS * 1.5;
+        font-weight: 700;
+      }
+      .lnsMetricExpression__value {
+        font-size: $euiSizeM * 2.25;
+      }
```
Why we are changing also the `font-weight` in this case?

markov00 commented on x-pack/plugins/lens/common/expressions/metric_chart/metric_chart.ts:
```diff
@@ -38,6 +38,24 @@ export const metricChart: ExpressionFunctionDefinition<
         defaultMessage: 'The chart title.',
       }),
     },
+    titleSize: {
+      types: ['string'],
+      help: i18n.translate('xpack.lens.metric.titleSize.help', {
+        defaultMessage: 'The chart title size.',
+      }),
+    },
```
From the CSS we are not just changing the size of the label but also the size of the value. We should clearly describe that here

MichaelMarcialis commented on x-pack/plugins/lens/public/metric_visualization/expression.scss:
```diff
@@ -11,11 +11,81 @@
+        font-size: $euiSizeS * 1.5;
+        font-weight: 700;
+      }
+      .lnsMetricExpression__value {
+        font-size: $euiSizeM * 2.25;
+      }
+    }
+    &.titleSizeM {
+      .lnsMetricExpression__title {
+        font-size: $euiSizeM * 1.5;
+      }
+      .lnsMetricExpression__value {
+        font-size: $euiSizeL * 2;
+      }
+    }
+    &.titleSizeL {
+      .lnsMetricExpression__title {
+        font-size: $euiSizeL;
+      }
+      .lnsMetricExpression__value {
+        font-size: $euiSizeXL * 2;
+      }
+    }
+    &.titleSizeXL {
+      .lnsMetricExpression__title {
+        font-size: $euiSizeXL;
+      }
+      .lnsMetricExpression__value {
+        font-size: $euiSizeXXL * 2;
+      }
+    }
+
+    &.titleSizeXXL {
+      .lnsMetricExpression__title {
+        font-size: $euiSizeXXL;
+      }
+      .lnsMetricExpression__value {
+        font-size: $euiSizeXXL * 3;
+      }
+    }
```
Rather than using the `$euiSize` series of variables, would it be better to use the `$euiFontSize` series of variables?

MichaelMarcialis commented on x-pack/plugins/lens/public/metric_visualization/metric_config_panel/title_aize_option.tsx:
```diff
@@ -0,0 +1,58 @@
+/*
```
This file is currently named `title_aize_option.tsx`. I believe it's meant to be `title_size_option.tsx`.

MichaelMarcialis commented on x-pack/plugins/lens/public/metric_visualization/metric_config_panel/title_aize_option.tsx:
```diff
@@ -0,0 +1,58 @@
+}
+
+const titleSizes = [
+  { id: 'xs', label: 'XS' },
+  { id: 's', label: 'S' },
+  { id: 'm', label: 'M' },
+  { id: 'l', label: 'L' },
+  { id: 'xl', label: 'XL' },
+  { id: 'xxl', label: 'XXL' },
```
Should we be using abbreviations for the different size options? Or should we be spelling them out fully for the users (ex. Extra small, Small, Medium, etc.). Since we have the space in the selector, I'm inclined to want them spelled fully. Thoughts?

MichaelMarcialis commented on x-pack/plugins/lens/public/metric_visualization/metric_config_panel/title_aize_option.tsx:
```diff
@@ -0,0 +1,58 @@
+      label={
+        <>
+          {i18n.translate('xpack.lens.metricChart.titleSizeLabel', {
+            defaultMessage: 'Title size',
```
Per previous comments, change this label if it will affect both metric title and value size.

MichaelMarcialis commented on x-pack/plugins/lens/public/metric_visualization/metric_config_panel/title_align_option.tsx:
```diff
@@ -0,0 +1,62 @@
+      label={
+        <>
+          {i18n.translate('xpack.lens.metricChart.titleAlignLabel', {
+            defaultMessage: 'Title align',
```
Per previous comments, change this label if it will affect both metric title and value alignment.

MichaelMarcialis commented on x-pack/plugins/lens/public/metric_visualization/metric_config_panel/title_align_option.tsx:
```diff
@@ -0,0 +1,62 @@
+        </>
+      }
+    >
+      <EuiSuperSelect
+        data-test-subj="lnsMissingValuesSelect"
+        compressed
+        options={titleAlignPositions
+          .filter((pos) => pos.direction !== direction)
+          .map((position) => {
+            return {
+              value: position.id,
+              dropdownDisplay: position.label,
+              inputDisplay: position.label,
+            };
+          })}
+        valueOfSelected={state.titleAlignPosition ?? 'center'}
+        onChange={(value) => {
+          setState({ ...state, titleAlignPosition: value });
+        }}
+        itemLayoutAlign="top"
+        hasDividers
+      />
```
It may be worth considering to change this `EuiSuperSelect` component to an `EuiButtonGroup`, to require one less click from users and also be more consistent with other areas of the Lens UI.

MichaelMarcialis commented on x-pack/plugins/lens/public/metric_visualization/metric_config_panel/title_position_option.tsx:
```diff
@@ -0,0 +1,56 @@
+        </>
+      }
+    >
+      <EuiSuperSelect
+        data-test-subj="lnsMissingValuesSelect"
+        compressed
+        options={titlePositions.map((position) => {
+          return {
+            value: position.id,
+            dropdownDisplay: position.label,
+            inputDisplay: position.label,
+          };
+        })}
+        valueOfSelected={state.titlePosition ?? 'top'}
+        onChange={(value) => {
+          setState({ ...state, titlePosition: value });
+        }}
+        itemLayoutAlign="top"
+        hasDividers
+      />
```
It may be worth considering to change this `EuiSuperSelect` component to an `EuiButtonGroup`, to require one less click from users and also be more consistent with other areas of the Lens UI.

MichaelMarcialis commented on x-pack/plugins/lens/public/metric_visualization/metric_config_panel/visual_options_popover.tsx:
```diff
@@ -0,0 +1,42 @@
+        buttonDataTestSubj="lnsVisualOptionsButton"
+      >
+        <TitleSizeOptions state={state} setState={setState} />
+        <EuiHorizontalRule />
```
Consider removing `EuiHorizontalRule`.

MichaelMarcialis commented on x-pack/plugins/lens/public/metric_visualization/metric_config_panel/visual_options_popover.tsx:
```diff
@@ -0,0 +1,42 @@
+          defaultMessage: 'Visual options',
+        })}
+        type="visualOptions"
+        groupPosition="left"
```
To avoid showing sharp corners on the right side of this toolbar button, please remove this `groupPosition` prop (as it's the only button in this group).

![image](https://user-images.githubusercontent.com/3884767/151877078-b2a58ed8-547a-497f-8889-78404654224e.png)

MichaelMarcialis reviewed (commented):
Thanks for putting this together, @shahzad31! Left you a few comments and questions for your review.

shahzad31 commented on x-pack/plugins/lens/common/expressions/metric_chart/metric_chart.ts:
```diff
@@ -38,6 +38,24 @@ export const metricChart: ExpressionFunctionDefinition<
+        defaultMessage: 'The chart title size.',
+      }),
+    },
+    titlePosition: {
+      types: ['string'],
+      help: i18n.translate('xpack.lens.metric.titlePosition.help', {
+        defaultMessage: 'The chart title position.',
+      }),
+    },
```
Position is actually relevant here i mean title is position relevant to value.

shahzad31 commented on x-pack/plugins/lens/public/metric_visualization/expression.scss:
```diff
@@ -11,11 +11,81 @@
+    display: flex;
+    align-items: center;
+    flex-direction: column;
+    &.rowDirection {
+      flex-direction: row;
+    }
```
I have removed left/right. I think we can add those if we found use case. I can't

shahzad31 commented:
Thanks @markov00 and @MichaelMarcialis i have address most of your feedback 

1. Converted Align select to Button groups
2. Remove left/right positions items
3. Using `EuiFontSize` now 
4. Since size/align apply to both, have remove title prefix from them
5. Removed `EuiHorizentalRule`
6. Updated class name to better reflect
7. Changed the title to `Appearance` 
8. 
![image](https://user-images.githubusercontent.com/3505601/151978838-1297c482-290c-404e-b427-5dbf494de7e3.png)

MichaelMarcialis reviewed (commented):
Thanks so much for addressing those comments, @shahzad31! After re-reviewing, I have a few additional comments.

Also, if it's cool with you, I took a quick design pass at the appearance toolbar menu (see image below). Changes include:

- Tightened up a bit of the vertical space by consolidating the size and alignment options to a single form row.
- Changed the title position select component to a button group.
- Provided new prepend/append buttons to increase/decrease the text size as a faster alternative (in addition to being able to open the text size select dropdown as they can currently). 

![image](https://user-images.githubusercontent.com/3884767/152055962-9368b78a-e4d1-44e8-8c07-99617f931418.png)

Any interest in making these changes as part of this PR? FYI, for the sake of space, I did end up changing the text size names back to abbreviations. Apologies for the back-and-forth on that!

shahzad31 commented:
@MichaelMarcialis Sure, i like the update, will push the changes.

markov00 commented on x-pack/plugins/lens/common/expressions/metric_chart/metric_chart.ts:
```diff
@@ -38,6 +38,24 @@ export const metricChart: ExpressionFunctionDefinition<
+      help: i18n.translate('xpack.lens.metric.titleAlignPosition.help', {
+        defaultMessage: 'The chart title align position.',
+      }),
+    },
```
@MichaelMarcialis I'm a bit confused because I don't see where in Lens you are using the words `title` and `subtitle` in the single metric visualizations. 
<img width="400" alt="Screenshot 2022-02-03 at 12 55 05" src="https://user-images.githubusercontent.com/1421091/152338897-be463754-b6df-45b1-b097-0bf129f2122a.png">
I see a `Display name` field that I can change to change the metric name, but I don't see any title or subtitle in any of our options. The only place I see the `title` word is in the CSS class name.
In the other metric vis (aggregated) I can see a `Custom label` that can be used to change the metric name.
<img width="400" alt="Screenshot 2022-02-03 at 12 56 22" src="https://user-images.githubusercontent.com/1421091/152339085-17e068a2-41b7-4478-9a14-5c52949f95e2.png">
This is actually badly aligned with the option to show and hide that name called `Show title`
<img width="400" alt="Screenshot 2022-02-03 at 12 56 17" src="https://user-images.githubusercontent.com/1421091/152339286-05ab37cb-b443-4093-adb8-a257f7bc4ac5.png">

The title is used on the `Axis` instead, where you can configure the `Axis title` in the top menu that changes the `Display name` of the vertical or horizontal axis.
<img width="381" alt="Screenshot 2022-02-03 at 13 12 39" src="https://user-images.githubusercontent.com/1421091/152340568-0ca1a1a5-570f-4709-b2c7-3bd7355b78a3.png">

When I think of a `title`, to me always refers to something more descriptive like: `Memory consumption in the last week` and is more suitable to be places on the top-left part of the visualization (aka panel title). A `label` instead, is used in various places and contains usually less information, just a tag: `RAM`, `CPU %`, really similar to what `name` is about.

As always is a matter of consistency across the interface and I don't see that neither with the term `title` or the `label`.
cc @gvnmagni

shahzad31 commented:
@MichaelMarcialis i have updated , there is one small issue when XXL is selected, First row bit overflows, i am not sure, if we need to increase popover width.

<img width="1190" alt="image" src="https://user-images.githubusercontent.com/3505601/152501062-a3494f27-a22f-4158-be0c-859968ea5b2b.png">

MichaelMarcialis reviewed (approved):
Hey, @shahzad31. Thanks so much for making those suggested changes. Apologies for my delayed response; I'm just returning from PTO.

This looks great. Upon re-review, the only issues I discovered were those related to the form fields overflowing outside of the popover, which you mentioned earlier. Below is a list of suggested changes that should resolve that issue. In addition to these changes, you'll need to also change the width of the popovers via the `toolbar_popover.scss` styles (as it's currently too narrow to house these updated form elements). I used `440px` in my quick mockups, so feel free to use that assuming it doesn't break anything.

Otherwise, I'm good to approve this, assuming those changes get made. Thanks so much again!

flash1293 commented on x-pack/plugins/lens/public/metric_visualization/auto_scale.tsx:
```diff
@@ -85,6 +90,12 @@ export class AutoScale extends React.Component<Props, State> {
+                alignLeft: textAlign === 'left',
+                alignRight: textAlign === 'right',
+                alignCenter: textAlign === 'center',
+                [`titleSize${(size ?? 'xl').toUpperCase()}`]: true,
```
Any reason why the default size isn't handled on the `toExpression` level like position and alignment?

VladLasitsa replied:
removed default here because we use default font size in class `lnsMetricExpression__value` and don't needed to add this class if size is not defained.

flash1293 commented on x-pack/plugins/lens/public/metric_visualization/visualization.tsx:
```diff
@@ -59,6 +60,9 @@ const toExpression = (
         arguments: {
           title: [attributes?.title || ''],
+          size: [state?.size || ''],
+          titlePosition: [state?.titlePosition || 'top'],
```
Not opposed to default to "top" for new visualizations, but the old default was "bottom". So if we want to keep the default value here we have to add a migration to add in the `bottom` for the existing ones.

flash1293 reviewed (commented):
Thank you so much for working on this @shahzad31 ! This will do wonders for Lens based dashboards.

I left two smallish issues, but otherwise it looks great to me

VladLasitsa commented:
@elasticmachine merge upstream

VladLasitsa commented:
> Also it seems like the default font size was slightly larger on old versions: <img alt="Screenshot 2022-02-11 at 10 25 44" width="302" src="https://user-images.githubusercontent.com/1508364/153566928-2fa0950d-2cbf-474f-a3e6-a3167bb54d67.png"> <img alt="Screenshot 2022-02-11 at 10 25 14" width="441" src="https://user-images.githubusercontent.com/1508364/153566935-7d705227-7dde-4f0d-87a3-cd33f79dc318.png">
> 
> Not a blocker for me but as FYI

It's because previously we used euiSize but now we are using euiFontSize. From my perspective using of euiFontSize is more correct.
![image](https://user-images.githubusercontent.com/16915480/154062034-548b0b70-4854-49e1-bbd9-c1fb81b6aa26.png)

ryankeairns commented:
Thanks for this @shahzad31 ! This will really help us in improving the visual appearance of dashboards.
